src/xrpld/rpc/handlers/DepositAuthorized.cpp

```diff
@@ -88,23 +90,111 @@ doDepositAuthorized(RPC::JsonContext& context)
         return result;
     }
 
-    // If the two accounts are the same, then the deposit should be fine.
-    bool depositAuthorized{true};
-    if (srcAcct != dstAcct)
+    bool const reqAuth =
+        (sleDest->getFlags() & lsfDepositAuth) && (srcAcct != dstAcct);
+    bool const credentialsPresent = params.isMember(jss::credentials);
+
+    std::set<std::pair<AccountID, Slice>> sorted;
+    std::vector<std::shared_ptr<SLE const>> lifeExtender;
+    if (credentialsPresent)
     {
-        // Check destination for the DepositAuth flag.  If that flag is
-        // not set then a deposit should be just fine.
-        if (sleDest->getFlags() & lsfDepositAuth)
+        auto const& creds(params[jss::credentials]);
+        if (!creds.isArray() || !creds)
+        {
+            return RPC::make_error(
+                rpcINVALID_PARAMS,
+                RPC::expected_field_message(
+                    jss::credentials,
+                    "is non-empty array of CredentialID(hash256)"));
+        }
+        else if (creds.size() > maxCredentialsArraySize)
+        {
+            return RPC::make_error(
+                rpcINVALID_PARAMS,
+                RPC::expected_field_message(
+                    jss::credentials, "array too long"));
+        }
+
+        lifeExtender.reserve(creds.size());
+        for (auto const& jo : creds)
         {
-            // See if a preauthorization entry is in the ledger.
-            auto const sleDepositAuth =
-                ledger->read(keylet::depositPreauth(dstAcct, srcAcct));
-            depositAuthorized = static_cast<bool>(sleDepositAuth);
+            if (!jo.isString())
+            {
+                return RPC::make_error(
+                    rpcINVALID_PARAMS,
+                    RPC::expected_field_message(
+                        jss::credentials, "an array of CredentialID(hash256)"));
+            }
+
+            uint256 credH;
+            auto const credS = jo.asString();
+            if (!credH.parseHex(credS))
+            {
+                return RPC::make_error(
+                    rpcINVALID_PARAMS,
+                    RPC::expected_field_message(
+                        jss::credentials, "an array of CredentialID(hash256)"));
+            }
+
+            std::shared_ptr<SLE const> sleCred =
+                ledger->read(keylet::credential(credH));
+            if (!sleCred)
+            {
+                RPC::inject_error(
```

Comment:
Not needed in this PR, just noting for the future: it'd be nice if the error message also told you which credential was the problem.